### DISCUSSION TITLE
Is there a way to add control panel/admin function to an OrchardCore app outside the CMS?

### DISCUSSION THREAD

**Discussion by MikeAlhayek** [OrchardCMS/OrchardCore#7716] [Q&A]
Hello,
I am researching best approach for a new project I am working on. Here are some of the features I am looking to have

1. Multi-tenant Saas Application
2. Modular 
3. A backed control panel to manage users, roles, claims and view reports.
4. A way to add items to a menu on the fly. For example, add items to the menu only if the module is available to the tenant.

I can simply use OrchardCore to achieve 1 & 2 very. However, is there more built-in modules/functionality that I can use to add a control panel with the app with a way to add items to the menu similar to the way the CMS app works? 

Additionally, is there a module that I can use to manage the users "reset password, change roles, etc..." or the roles/claims a user can have when he/she logs in?

If these common function already exists, I Would appreciate an example of how to setup the control panel with the menu.

**Comment by Skrypt** [comment]
I think that the SAAS recipe is probably what you are looking for. Though, maybe you need even less features to be enabled by default but the SAAS recipe is definitely a good starting point.

**Reply by MikeAlhayek** [reply]
doesn't recipes require the use of the CMS? if not, can you please explain more what is saas recipe?

**Reply by Skrypt** [reply]
I can't say, I would need to analyze this and come back with a clear answer on this. Of course, you can use the OrchardCore abstractions and implement your own things with them. But that's not the goal in the end right?

Now, I think what you are asking for is something that we need to refactor in OC. We had a discussion about this in the last 3 weeks and we need to come up with a better approach on the "framework" aspect of things.

Of course I suggested to you using the SAAS recipe because it simply enables the basic features of Orchard Core but it also requires you to use the CMS admin. As long as you are not enabling Content Types module you won't really enable the "CMS dynamic bits" of Orchard Core.

**Reply by MikeAlhayek** [reply]
I don't mind to test out the out come of the Saas recipe and see if it'll do or not. How can I add the saas recipe to my existing project? I already have `OrchardCore.Application.Mvc.Targets`  installed. What else do I need to install to get the SaaS recipe in place?

**Reply by Skrypt** [reply]
The SaaS recipe can be executed when using the setup of a tenant. Else in the admin you can execute recipes too. If you follow the documentation on how to get OrchardCore running with a Nuget Package solution you should be fine. Look at the templates we have for `dotnet new`.

https://github.com/OrchardCMS/OrchardCore/tree/dev/src/Templates/OrchardCore.ProjectTemplates/content

**Reply by MikeAlhayek** [reply]
@Skrypt I was able to run a saas recipe. Yes this is all the function that I need. However, this force me to use YesSql which I can’t use in this application. Is there a way to implement my own stores so I can use this saas recipe but store the data differently? I think if there is a data abstraction layer available, I can implement my own stores. My project already use UnitOfWork pattern to store data and I like to stick with my UnitOfWork.

**Reply by MikeAlhayek** [reply]
Here is my UnitOfWork. As you can see I added identity repositories + extend the security features of identity https://github.com/CrestApps/CrestApps.Core/blob/master/CrestApps.Data.Abstraction/IUnitOfWork.cs

**Reply by Skrypt** [reply]
So, in which context you want to use this UnitOfWork? Because it could be used inside an Orchard Core module. There's nothing preventing you using your own DAL from inside an Orchard Core module. The CMS though requires YesSQL because of the dynamic nature of Content Types in OC. But you could, not use them at all in this case. For the other modules you will need to look if they will get applied in the context of a OC tenant. Please take a look around here on Github for answers given by @jtkech about the tenant feature.

**Reply by MikeAlhayek** [reply]
@Skrypt, I am able to use the my unit-of-work in my modules with no problem. I was hoping to use my UnitOfWork in place of YesSQL which does not seem possible. I extended the identity security to allow me to prevent the use of the same password in a 12 month period. I also extended the lockout function to have a lockout count for the lockout. For example, if I user got locked out 5 times back to back, I want to lockout the user for 6 months instead of 5 mins "more like a long ban" is there a way to implement the same with the Saas recipe? Also, I would want to add DefaultlocationId to the user profile, how can I add new field to the user?

**Reply by Skrypt** [reply]
There are some few points here that I understand and that could probably work but I need to make more extensive tests. And I believe that we are now in that period of time where these things will be asked a lot about tenants and how to use them without YesSQL. We probably can't as of now but there's surely either some development that could be done on tenants or things I don't know yet. @jtkech is the one that did the tenants feature I will ask him when I get time or he will answer us here. But, I believe that this is absolutely something that can be done.

**Comment by jtkech** [comment]
@CrestApps the `OrchardCore` framework is not coupled to `YesSql` but `OC.Infrastructure` is, around half of the modules are not coupled to `YesSql` but around half of them are coupled. We have some documents managed by an `IDocumentStore` that can be overridden, but as said by @Skrypt content items management is coupled to `YesSql`, you would need to rewrite e.g. `OC.ContentManagement`, `OC.Contents` and `OC.ContentTypes` modules, and modules that depend on them.

`OC.Infrastructure` => `OC.Data` => `YesSql`
Module `OC.Contents` => `OC.ContentManagement` => `OC.Infrastructure`

> I would want to add DefaultlocationId to the user profile, how can I add new field to the user?

We already have custom user settings that can be used to define different user profiles
But `OC.Users` => `OC.Infrastructure` => `OC.Data` => `YesSql`

**Comment by Skrypt** [comment]
Just as a follow up. Last night we tried creating multiple tenants by using a combination of DI and appsettings.json file to see if we could create tenants that use Orchard Core and other that are simple MVC applications. It works partially. We need to test more but it should definitely work as the code is all there. The only thing that doesn't work is when we are adding an Orchard Core tenant with other simple MVC tenants. Need to investigate further to see why it doesn't work. The tenant seems to be created but the / route of the tenant returns a 404 page not found.

So good news, we can create tenants that are not Orchard Core tenants.

Here a solution for those who wants to keep using their MVC custom app would be to create a tenant with the assemblies of this MVC app and a second tenant with Orchard Core accessible with a "urlprefix". This would allow us to migrate and test progressively each modules that could potentially be used standalone inside a MVC app this way.

**Comment by hishamco** [comment]
@jtkech seems `OrchardCore.Data.Abstractions` has a dependency on YesSQL which isn't expected :( I though `Orchard.Data` has a dependency which it's a YesSQL implementation

IMHO Orchard Core as Framework shouldn't depend on YesSQL but Orchard Core CMS could, your thought @sebastienros

**Reply by MikeAlhayek** [reply]
Yes, that is what I run into. This is the issue I am facing now and is preventing me from using OrchardCore for some apps that I would love to convert to Orchard. 

IMHO, the storage for the core function should be flexible and changed as needed. Having the default implementation to be YesSQL is great for most users. But in some cases, "my case", I need to be able to change the default implementation for some of the core function like "users and Identity related tables". In my case, the `users` and the `roles` table is used by many micro-services some using apps and some using sql directly. Storing everything into the document table is simply not an option. I feel this would also be true from many enterprise applications too. For content managements, storing everything is YesSQL is fine but the core function should allow for swapping out the implementation.

**Reply by TFleury** [reply]
I opened an issue (#9245 ) about `OrchardCore.Data.Abstractions` dependency on `YesSql`.

**Comment by jtkech** [comment]
So, one solution in your case would be to implement a module (or from the app with our helpers) this to override our `Usertore` and `RoleStore`. The module would need to depend on `OC.Users` and `OC.Roles` so that its registrations are done after and override the previous ones, or with an higher order property in the module startup (or from the app with our helpers with an higher order parameter).

**Reply by MikeAlhayek** [reply]
I am just returning to attempt this but running into issue #8060

**Reply by jtkech** [reply]
I meant through the DI re-register all or part of the following

            services.TryAddScoped<UserStore>();
            services.TryAddScoped<IUserStore<IUser>>(sp => sp.GetRequiredService<UserStore>());
            services.TryAddScoped<IUserRoleStore<IUser>>(sp => sp.GetRequiredService<UserStore>());
            services.TryAddScoped<IUserPasswordStore<IUser>>(sp => sp.GetRequiredService<UserStore>());
            services.TryAddScoped<IUserEmailStore<IUser>>(sp => sp.GetRequiredService<UserStore>());
            services.TryAddScoped<IUserSecurityStampStore<IUser>>(sp => sp.GetRequiredService<UserStore>());
            services.TryAddScoped<IUserLoginStore<IUser>>(sp => sp.GetRequiredService<UserStore>());
            services.TryAddScoped<IUserClaimStore<IUser>>(sp => sp.GetRequiredService<UserStore>());
            services.TryAddScoped<IUserAuthenticationTokenStore<IUser>>(sp => sp.GetRequiredService<UserStore>());

But by using `AddScoped<>()` not `TryAddScoped<>()`, this through a module that depends on `OC.Users` so that they are register after, hmm but here if your module startup run before, it may work as above `OC.Users` uses `TryAddScoped`.

**Reply by MikeAlhayek** [reply]
@jtkech sorry I am just getting back to this. If I understand your recommendation correctly, your asking me to create my own store like `CustomUserStore : IUserStore<IUser>` then have my own implementation of that interface which will be very similar to orchard implementation, except when the data is stored. Correct? 

Assuming this is correct, making orchard UserStore methods virtual will make my implementation much cleaner and I won't have to redefine the same logic. If I can override the `UserStore`, all I'll need to do is override few methods not the whole thing. If you look at my [PR](https://github.com/OrchardCMS/OrchardCore/pull/8061) you'll see that I wrapped the methods that write to the database into methods which I'll override.

To be exact, all I'll need to override are `SaveUserAsync(newUser, true)`,  `GenerateUniqueUserIdAsync` and `DeleteUser(user, true)``. and If I want to keep using a random string for id, I'll be able to do that by overriding `GetUserCountAsync` instead of `GenerateUniqueUserIdAsync`

**Reply by MikeAlhayek** [reply]
@jtkech sorry I am just getting back to this. If I understand your recommendation correctly, you're asking me to create my own store like `CustomUserStore` which implements `IUserStore<IUser>` then have my own implementation of that interface; which will be very similar to orchard implementation, except how the data is stored. Correct? 

Assuming this is correct, making orchard UserStore methods virtual will make my implementation much cleaner and manageable since I won't have to redefine the same logic. If I can override the methods from the `UserStore`, all I'll need to do is override few methods not the whole thing. If you look at my [PR](https://github.com/OrchardCMS/OrchardCore/pull/8061) you'll see that I wrapped the methods that write to the database into methods which I'll override.

To be exact, with my [PR](https://github.com/OrchardCMS/OrchardCore/pull/8061), all I'll need to override are `SaveUserAsync(newUser, true)`,  `GenerateUniqueUserIdAsync` and `DeleteUser(user, true)`. Alternately, if I want to keep using a random string for id using `IUserIdGenerator`, I'll be able to do that by overriding `GetUserCountAsync` instead of `GenerateUniqueUserIdAsync`

**Reply by jtkech** [reply]
So you can just override `IUserStore<IUser>`

Currently registered like this

            services.TryAddScoped<UserStore>();
            services.TryAddScoped<IUserStore<IUser>>(sp => sp.GetRequiredService<UserStore>());

**Reply by MikeAlhayek** [reply]
yes that is the plan. So are you saying that making the UserStore methods are virtual is not a good idea as per the PR request?

**Reply by jtkech** [reply]
Yes if you just need to override `IUserStore`, not the others `ISomethingStore`

        services.AddScoped<IUserStore, YourUserStor>();

**Reply by xperiandri** [reply]
@CrestApps were you successful in making the admin panel work on top of Orchard Core Framework?

**Reply by MikeAlhayek** [reply]
@xperiandri unfortunately I did not have the time to do it as the project I was working on was paused!

**Comment by TFleury** [comment]
This discission is quite old, but I'm also interested by this scenario.
I'm trying to isolate CMS bytes and only keep multi-tenant management and Admin UI.

My idea is also to replace document database (YesSql) with a relational database (EF Core)

Could it be possible to add Saas sample project to share work on this ?

**Reply by MikeAlhayek** [reply]
@TFleury in theory you should be able to use the core module outside of the CMS. However, I tried it one time and encountered few issue "some have been addresses in v1.5"

If you have the time, you may want attempt that path and see what kind of issues you'll encounter.

**Comment by TFleury** [comment]
The big part is to remove direct dependencies on YesSql.
I already made 2 PRs to address this (#12826 and #12906)

The next step will probably be the User and Role stores.

**Comment by TFleury** [comment]
I got it working and I've proposed to include it as a sample. (OrchardCMS/OrchardCore.Samples#35)
But there is still work to allow replacing document database with other providers.

**Reply by sambeauvois** [reply]
Awesome, thanks for this sample.
I was looking for a way to use orchard core framework parts only and besides the modular system I think there is not a lot of examples.
